core/client/fs/src/main/java/alluxio/client/block/stream/SharedGrpcDataReader.java

```diff
@@ -0,0 +1,192 @@
+    mBlockId = readRequest.getBlockId();
+    mCachedDataReader = reader;
+  }
+
+  @Override
+  public long pos() {
+    return mPosToRead;
+  }
+
+  /**
+   * Seeks to a specific position.
+   *
+   * @param pos the position to seek to
+   */
+  public void seek(long pos) {
+    mPosToRead = pos;
+  }
+
+  @Override
+  @Nullable
+  public DataBuffer readChunk() throws IOException {
+    int index = (int) (mPosToRead / mChunkSize);
+    DataBuffer chunk = mCachedDataReader.readChunk(index);
+    if (chunk == null) {
+      return null;
+    }
+    ByteBuffer bb = chunk.getReadOnlyByteBuffer();
+    // Force to align to chunk size
+    bb.position((int) (mPosToRead % mChunkSize));
+    mPosToRead += mChunkSize - mPosToRead % mChunkSize;
+
+    return new NioDataBuffer(bb, bb.remaining());
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mCachedDataReader.deRef() > 0) {
+      return;
+    }
+    try (LockResource lockResource = new LockResource(getLock(mBlockId).writeLock())) {
+      if (mCachedDataReader.getRefCount() == 0) {
+        BLOCK_READERS.remove(mBlockId);
+      }
+    }
+    if (mCachedDataReader.getRefCount() == 0) {
+      mCachedDataReader.close();
+    }
+  }
+
+  /**
+   * Factory class to create {@link SharedGrpcDataReader}s.
+   */
+  public static class Factory implements DataReader.Factory {
+    private final FileSystemContext mContext;
+    private final WorkerNetAddress mAddress;
+    private final ReadRequest mReadRequestPartial;
+    private final long mBlockSize;
+
+    /**
+     * Creates an instance of {@link SharedGrpcDataReader.Factory} for block reads.
+     *
+     * @param context the file system context
+     * @param address the worker address
+     * @param readRequestPartial the partial read request
+     * @param blockSize the block size
+     */
+    public Factory(FileSystemContext context, WorkerNetAddress address,
+        ReadRequest readRequestPartial, long blockSize) {
+      mContext = context;
+      mAddress = address;
+      mReadRequestPartial = readRequestPartial;
+      mBlockSize = blockSize;
+    }
+
+    @edu.umd.cs.findbugs.annotations.SuppressFBWarnings(
+        value = "AT_OPERATION_SEQUENCE_ON_CONCURRENT_ABSTRACTION",
+        justification = "operation is still atomic guarded by block Ølock")
+    @Override
+    public DataReader create(long offset, long len) throws IOException {
+      long blockId = mReadRequestPartial.getBlockId();
+      BufferCachingGrpcDataReader reader;
+      try (LockResource lockResource = new LockResource(getLock(blockId).writeLock())) {
+        reader = BLOCK_READERS.get(blockId);
+        if (reader == null) {
+          // Even we may only need a portion, create a reader to read the whole block
+          ReadRequest cacheRequest = mReadRequestPartial
+              .toBuilder().setOffset(0).setLength(mBlockSize).build();
+          reader = BufferCachingGrpcDataReader.create(mContext, mAddress, cacheRequest);
```

Comment:
Is this `SharedGrpcDataReader` the only class that will create a `BufferCachingGrpcDataReader` instance?

Reply:
yes